secret.go

```diff
@@ -3,27 +3,45 @@ package secretcrypt
 import (
 	"fmt"
 	"strings"
+	"sync"
 
 	"github.com/Zemanta/go-secretcrypt/internal"
 )
 
 // Secret represents an encrypted secret.
 type Secret struct {
+	lock          sync.RWMutex
 	crypter       internal.Crypter
 	ciphertext    internal.Ciphertext
+	plaintext     string
+	cached        bool
 	decryptParams internal.DecryptParams
 }
 
 // Decrypt decrypts the secret and returns the plaintext.
-func (s Secret) Decrypt() (string, error) {
+func (s *Secret) Decrypt() (string, error) {
 	if s.crypter == nil || s.ciphertext == "" {
 		return "", fmt.Errorf("Cannot decrypt a zero secret!")
 	}
-	return s.crypter.Decrypt(s.ciphertext, s.decryptParams)
+	s.lock.RLock()
```

Comment:
Not true with Go's memory model; it can happen that s.cached==true even if s.plaintext isn't set yet when they are being set concurrently.

"Note that a read r may observe the value written by a write w that happens concurrently with r. Even if this occurs, it does not imply that reads happening after r will observe writes that happened before w."
- from https://golang.org/ref/mem

Which reminds me, it'd be better to be idiomatic and use sync.Once here :-D